### ISSUE TITLE
About function "transform" in transforms.py

### ISSUE THREAD

**Issue by zhiqiangdon** [bearpaw/pytorch-pose#19]
Hi,

Thanks for your code. I have one question about some code in the "transform" function. 

 new_pt = np.array([pt[0] - 1, pt[1] - 1, 1.]).T
 new_pt = np.dot(t, new_pt)
 return new_pt[:2].astype(int) + 1

According to the above code, you first subtract 1 from the coordinates and then add 1 after the transformation. I don't see the reason of doing this. There are two places calling this "transform" function. The first place is in datasets/mpii.py function, 

 tpts[i, 0:2] = to_torch(transform(tpts[i, 0:2]+1, c, s, [self.out_res, self.out_res], rot=r))
 target[i] = draw_labelmap(target[i], tpts[i]-1, self.sigma, type=self.label_type)

Here you first add 1 and then subtract 1 before and after calling the "transform" function, which just offset what you do inside it. For this case, we could remove the plus 1 and minus 1 for clarity.

Second, function "final_preds" calls function "transform_preds" which then calls "transform" as follows:

coords[p, 0:2] = to_torch(transform(coords[p, 0:2], center, scale, res, 1, 0))

In this case, I also read the original torch code:
https://github.com/anewell/pose-hg-demo/blob/master/util.lua
It seems they don't add 1 and subtract 1 afterwards. I think adding 1 is not equivalent to subtracting 1 after the trasformation. Could please explain your reason?

Thanks,

**Comment by bestlin** [comment]
I think that is because the python and lua.

Thanks.

Lin

2017-09-25 14:23 GMT-07:00 zhiqiangdon <notifications@github.com>:

**Comment by zhiqiangdon** [comment]
Hi @bestlin ,

I know that python's index starts from 0 while lua index starts from 1. I don't think this is the reason. Because the index difference should considered in the data preprocessing part, i.e. generating the json files. A more important reason is that minus 1 in the resolution of 64x64 and plus 1 of 256x256 before is obviously not equivalent.

**Comment by bearpaw** [comment]
For `transform`, please refer to the original lua code. 
https://github.com/anewell/pose-hg-train/blob/master/src/util/img.lua#L54-L65
```lua
function transform(pt, center, scale, rot, res, invert)
    local pt_ = torch.ones(3)
    pt_[1],pt_[2] = pt[1]-1,pt[2]-1

    local t = getTransform(center, scale, rot, res)
    if invert then
        t = torch.inverse(t)
    end
    local new_point = (t*pt_):sub(1,2)

    return new_point:int():add(1)
end
```

The `+1` and `-1` in the following code is left here for historical reason. To compare with previous trained models, I will leave them unchanged.
```python
tpts[i, 0:2] = to_torch(transform(tpts[i, 0:2]+1, c, s, [self.out_res, self.out_res], rot=r))
target[i] = draw_labelmap(target[i], tpts[i]-1, self.sigma, type=self.label_type)
```

**Comment by zhiqiangdon** [comment]
@bearpaw ,Thanks! I was reading the file in repo pose-hg-demo:
https://github.com/anewell/pose-hg-demo/blob/master/img.lua

function transform(pt, center, scale, rot, res, invert)
    -- For managing coordinate transformations between the original image space
    -- and the heatmap

    local pt_ = torch.ones(3)
    pt_[1] = pt[1]
    pt_[2] = pt[2]
    local t = getTransform(center, scale, rot, res)
    if invert then
        t = torch.inverse(t)
    end
    local new_point = (t*pt_):sub(1,2):int()
    return new_point
end

Do you know why doesn't this transform function have the +1 and -1? Thanks!

**Comment by bearpaw** [comment]
@zhiqiangdon It seems that there are some mismatches between `pose-hg-demo` and `pose-hg-train`. I think `pose-hg-train` is relatively up-to-date.

I do not quite know the reason. But I guess it might due to the `int` operation. The trick might improve the computational accuracy (float and int stuff) and the performance a little bit. But it's absolutely not a key point, and I've never verified this either.

**Comment by zhiqiangdon** [comment]
@bearpaw Thanks!

**Comment by sydney0zq** [comment]
I have some questions about `transform.py`, can anyone help me out?
```
def get_transform(center, scale, res, rot=0):
    # General image processing matrix
    h = 200 * scale     # what does 200 means?
    t = np.zeros((3, 3)) # this is transform matrix, what is our aim? Any mathematical formula?
    t[0, 0] = float(res[1]) / h
    t[1, 1] = float(res[0]) / h
    t[0, 2] = res[1] * (-float(center[0]) / h + .5)
    t[1, 2] = res[0] * (-float(center[1]) / h + .5)
    t[2, 2] = 1   
```

Thanks, I just get stuck on them.

**Comment by xiaoyong** [comment]
Affine transform should be performed on 0-indexed points. So it's totally fine to remove the +1/-1 in `transform` and use consistent (0-indexed) coordinates.

@bearpaw There might be an inconsistence in

**Comment by bearpaw** [comment]
@sydney0zq This is the matrix for the affine transform. You may learn the relevant knowledge from image processing or computer vision courses. 

@xiaoyong Actually I am a little bit confused about this 1 to 0 index from lua to torch (and from the annotation). Would you create a pull request to clarify this if it won't cost much of your time? Thank you very much!

**Comment by xiaoyong** [comment]
@bearpaw Sure, I converted 1 to 0 index and everything works well, both on your pretrained model and my own retrained model. But I need to do some code tidy up. Maybe after the forthcoming holiday :-)

**Comment by zouxuelian** [comment]
Hello, thank you for contributing your code, your code is fine, but I have some questions about the lsp dataset.
I ran your code on the lsp dataset. The joint points of the training input are 16. The test generates predic_valid.mat is 1000*16*2. I checked the lsp real tag and found that the lsp dataset is 1000*14*2. How do I evaluate the lsp dataset on pck and how to remove the two extra points.
Thank you for your answer
